### PR TITLE
Increase resilience of the storage against data corruption.

### PR DESCRIPTION
**Pull request by beorn7**
@juliusv 

Nicely packed in semantically meaningful commits.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -59,8 +59,10 @@ var (
 
 	checkpointInterval         = flag.Duration("storage.local.checkpoint-interval", 5*time.Minute, "The period at which the in-memory metrics and the chunks not yet persisted to series files are checkpointed.")
 	checkpointDirtySeriesLimit = flag.Int("storage.local.checkpoint-dirty-series-limit", 5000, "If approx. that many time series are in a state that would require a recovery operation after a crash, a checkpoint is triggered, even if the checkpoint interval hasn't passed yet. A recovery operation requires a disk seek. The default limit intends to keep the recovery time below 1min even on spinning disks. With SSD, recovery is much faster, so you might want to increase this value in that case to avoid overly frequent checkpoints.")
+	seriesSyncStrategy         = flag.String("storage.local.series-sync-strategy", "adaptive", "When to sync series files after modification. Possible values: 'never', 'always', 'adaptive'. Sync'ing slows down storage performance but reduces the risk of data loss in case of an OS crash. With the 'adaptive' strategy, series files are sync'd for as long as the storage is not too much behind on chunk persistence.")
```
Comment (juliusv) [inline main.go]: https://www.google.de/search?q=define:synced

"syncing" and "synced" are the common ways of spelling it, without "'".

File: storage/local/persistence.go [hunk 2]
```diff
@@ -477,7 +482,11 @@ func (p *persistence) loadChunkDescs(fp clientmodel.Fingerprint, beforeTime clie
 //
 // (4.4) The varint-encoded persistWatermark. (Missing in v1.)
 //
-// (4.5) The varint-encoded chunkDescsOffset.
+// (4.5) The modification time of the series file as nanoceconds elapsed since
```
Comment (juliusv) [inline storage/local/persistence.go]: s/nanoceconds/nanoseconds/

File: storage/local/persistence.go [hunk 3]
```diff
@@ -1195,6 +1235,19 @@ func (p *persistence) openChunkFileForWriting(fp clientmodel.Fingerprint) (*os.F
 	// would still be detected.
 }
 
+// closeChunkFile first sync's the provided file if mandated so by the sync
```
Comment (juliusv) [inline storage/local/persistence.go]: sync's -> syncs

File: storage/local/persistence.go [hunk 4]
```diff
@@ -1195,6 +1235,19 @@ func (p *persistence) openChunkFileForWriting(fp clientmodel.Fingerprint) (*os.F
...
+// strategy. Then it closes the file. Errors are logged.
+func (p *persistence) closeChunkFile(f *os.File) {
+	if p.shouldSync() {
+		if err := f.Sync(); err != nil {
+			glog.Error("Error sync'ing file:", err)
```
Comment (juliusv) [inline storage/local/persistence.go]: dito etc.

File: storage/local/storage.go [hunk 5]
```diff
@@ -56,6 +62,21 @@ type evictRequest struct {
...
+
+// Possible values for SyncStrategy.
+const (
+	_ SyncStrategy = iota
+	Never
+	Always
+	Adaptive
+)
+
+// A syncStrategy is a function that returns if series files should be sync'd or
```
Comment (juliusv) [inline storage/local/storage.go]: if -> whether is better in this situation (avoids ambiguity)

File: storage/local/storage.go [hunk 6]
```diff
@@ -883,6 +933,23 @@ func (s *memorySeriesStorage) incNumChunksToPersist(by int) {
 	atomic.AddInt64(&s.numChunksToPersist, int64(by))
 }
 
+// isDegraded returns whether the storage is in "graceful degradation mode",
+// which is the case if the number of chunks waiting for persistence has reached
+// a percentage of maxChunksToPersist that exceepds
```
Comment (juliusv) [inline storage/local/storage.go]: exceepds -> exceeds

File: storage/local/storage.go [hunk 7]
```diff
@@ -883,6 +933,23 @@ func (s *memorySeriesStorage) incNumChunksToPersist(by int) {
...
+// percentChunksToPersistForDegradation. The method is not goroutine safe (but
+// only ever called from the goroutine dealing with series maintenance).
+// Changes of degradation mode are logged.
+func (s *memorySeriesStorage) isDegraded() bool {
+	nowDegraded := s.getNumChunksToPersist() > s.maxChunksToPersist*percentChunksToPersistForDegradation/100
+	if s.degraded && !nowDegraded {
+		glog.Warning("Storage has left graceful degradation mode.")
+	} else if !s.degraded && nowDegraded {
+		glog.Warning("%d chunks waiting for persistence (allowed maximum %d). Storage is now in graceful degradation mode. Series files are not sync'd anymore. Checkpoints will not be performed more often then every %v.", s.getNumChunksToPersist, s.maxChunksToPersist, s.checkpointInterval)
```
Comment (juliusv) [inline storage/local/storage.go]: then -> than

**Comment by juliusv** [comment]
:+1:

File: main.go (hunk 1 above)
Comment (beorn7) [inline main.go]: Done.

File: storage/local/persistence.go (hunk 2 above)
Comment (beorn7) [inline storage/local/persistence.go]: Done.

File: storage/local/persistence.go (hunk 3 above)
Comment (beorn7) [inline storage/local/persistence.go]: Done.

File: storage/local/persistence.go (hunk 4 above)
Comment (beorn7) [inline storage/local/persistence.go]: Done.

File: storage/local/storage.go (hunk 5 above)
Comment (beorn7) [inline storage/local/storage.go]: Done.

File: storage/local/storage.go (hunk 6 above)
Comment (beorn7) [inline storage/local/storage.go]: Done.

**Pull request merged**